src/ComponentFieldDefinition/FieldDefinition.php

```diff
@@ -0,0 +1,82 @@
+<?php declare(strict_types=1);
+
+namespace Torr\Storyblok\ComponentFieldDefinition;
+
+use Torr\Storyblok\Api\Manager\ComponentManager;
+use Torr\Storyblok\Data\DataSerializationContext;
+
+abstract class FieldDefinition
+{
+	public function __construct (
+		private readonly ?string $label = null,
+		private readonly ?int $position = null,
+		private readonly mixed $defaultValue = null,
+		private readonly ?string $description = null,
+		private readonly bool $translatable = false,
+		private readonly bool $required = false,
+		private readonly ?string $regexp = null,
+		private readonly ?bool $canSync = null,
+		private readonly ?bool $isPreviewField = null,
+	) {}
+
+	//region Field Accessors
+	public function getPosition () : ?int
+	{
+		return $this->position;
+	}
+
+	public function getDescription () : ?string
+	{
+		return $this->description;
+	}
+
+	public function isTranslatable () : bool
+	{
+		return $this->translatable;
+	}
+
+	public function isRequired () : bool
+	{
+		return $this->required;
+	}
+
+	public function getRegexp () : ?string
+	{
+		return $this->regexp;
+	}
+
+	public function isCanAsync () : ?bool
+	{
+		return $this->canSync;
+	}
+
+	public function isPreviewField () : ?bool
+	{
+		return $this->isPreviewField;
+	}
+	//endregion
+
+	abstract public static function getType () : FieldType;
+
+	/**
+	 * Returns the API representation of the custom fields of this Component Field that differ from the base FieldDefinition
+	 */
+	abstract public function getSchemaDefinition (ComponentManager $componentManager) : array;
```

Comment:
They can just overwrite the method? (it should not be final for that)

Reply:
From an API standpoint, I liked having just to implement the more specific data and not risking someone touching the base implementation. Same for having to constantly call `$serialized = parent::getSchemaDefinition($componentManager);` and then adding your custom stuff to `$serialized`. Though I'm definitely open for suggestions if we prefer this method.